File: packages/engine/src/plugins/backlogShed.ts
import type { BacklogShedConfig } from './PluginModule.js';

/**
 * Backlog shedding — the time-sync contract's latency RATCHET guard (ADR-0005).
 *
 * THE RATCHET. Under the contract every presentation consumer runs `sync=true`,
 * so its sink drains at exactly media rate. Backlog is then ONE-WAY: whatever
 * the leg's leaky queues absorb during a downstream hiccup (decode stall,
 * compositor hitch, CMA allocation) is never handed back, because a media-rate
 * sink has no spare rate to hand it back with. The legacy `sync=false` sink
 * presented on arrival and therefore gulped any backlog at max speed — which is
 * exactly the assumption the video leg's queue sizing was written against and
 * the contract falsified. Retained latency ratchets up one hiccup at a time
 * until the sink's own lateness handling (QoS upstream, then `max-lateness`)
 * costs nearly every frame. Field, .42 over ~16 h: 50 fps decoded, 2.5 fps on
 * the glass; a live +1 s `ts-offset` restored it instantly and reverting put it
 * straight back.
 *
 * WHY THIS LIVES AT THE CONTRACT LAYER. It is not a video-player problem and
 * not a queue-sizing problem — it is a property of every `sync=true` bus
 * consumer the contract creates (video-player, audio-decoder, and any future
 * presentation leg). So the knobs, the defaults and the gate on
 * `services.timeSyncContract` are defined ONCE here, exactly as
 * `effectivePlayoutOffsetMs` defines D once for both legs of a route. A leg
 * supplies only the two element names that are genuinely its own: where to
 * measure/shed, and which sink carries the route's `ts-offset`.
 *
 * WHAT IT IS NOT. Not a queue resize — the ES/jitter queues stay exactly as
 * they are, because their depth is field-measured IDR-burst absorption and
 * shrinking them re-breaks the picture. Not a bigger `ts-offset` — that is
 * unbounded and would desync the other leg of the same route. It is an ACTIVE
 * one-shot: detect retained latency past D, drop the OLDEST queued data (up to
 * the next keyframe on a video leg) until the leg is back at D, log it, and
 * stay quiet for a cooldown.
 *
 * The runner half is `_start_backlog_shedder` in `gst-pipeline-runner.py`
 * (measurement and dropping) plus `backlog_shed.py` (the sustained-excess and
 * rate-limit policy). The values below are the single source of truth for the
 * numbers; the runner's own fallbacks only exist so an older config still runs.
 */

/**
 * Plugin-event channel a shed episode is reported on. Pinned here because both
 * ends of a process boundary use it: the runner emits it, `GstPluginBase` logs
 * every leg's episodes off it.
 */
export const BACKLOG_SHED_EVENT = 'backlog_shed';

/**
 * Retained latency over D, in ms, that a leg may hold before it is shed. Sized
 * against the video leg's own queues: the ES queue holds 1 s and the jitter
 * queue `bufferMs` (200 ms default), so 250 ms is clear of a single absorbed
 * IDR burst (~200 ms of stream time at 8 Mbps on a Pi 4) and far below
 * `waylandsink max-lateness=1000000000`, the cliff past which the sink drops
 * nearly every frame.
 */
export const BACKLOG_SHED_TOLERANCE_MS = 250;

/**
 * How long the excess must hold before it counts. Every sample in the window
 * must be above tolerance (a floor, not an average — a spike relaxes, retained
 * latency does not), so a burst being absorbed and handed back cannot trip it.
 */
export const BACKLOG_SHED_HOLD_MS = 5_000;

/**
 * Minimum gap between sheds, per leg. A shed returns ALL of the excess, so
 * frequency buys nothing, and on a video leg each one costs the frames up to
 * the next IRAP — which the operator sees. One a minute, worst case.
 */
export const BACKLOG_SHED_COOLDOWN_MS = 60_000;

/**
 * Lateness past which a reading is NOT treated as a backlog. A real retained
 * backlog is bounded by the leg's queues (1 s ES + up to 5 s jitter); tens of
 * seconds means the buffer timeline and the pipeline clock are not the same
 * timeline, and shedding on that would drop the whole stream chasing a target
 * it can never reach. Reported, never acted on. Deliberately the same 10 s as
 * `MAX_PLAYOUT_OFFSET_MS` — no legitimate playout budget is larger.
 */
export const BACKLOG_SHED_SANITY_MS = 10_000;

/** The slice of `ModuleServices` the gate reads. */
export interface BacklogShedServices {
    timeSyncContract?: boolean;
}

export interface BacklogShedOptions {
    /**
     * `name=` of the element whose SINK pad is measured and shed at.
     *
     * Video: the DECODER. The backlog sits upstream of it (the ES and jitter
     * queues are what absorbed the stall), compressed AUs drop at I/O speed
     * where decoded frames would drain no faster than the decoder runs, and
     * `h26xparse` has already marked every access unit `DELTA_UNIT` or not —
     * which is what makes keyframe alignment possible at all.
     *
     * Audio: the SINK itself. Raw PCM references nothing, so the drop is safe
     * anywhere, and the sink's own pad is the last point that still sees every
     * buffer.
     */
    element: string;
    /**
     * `name=` of the presentation sink. Read (live) for its `ts-offset`, which
     * IS the route's playout offset D — so the measurement is against the
     * route's real budget including any operator trim, and a live D change is
     * picked up without rebuilding anything.
     */
    sink: string;
    /**
     * End the shed only on a keyframe. MANDATORY on a video leg feeding a
     * decoder: handing a stateless V4L2 decoder a delta unit whose references
     * were dropped is the documented hardware wedge (the same hazard the
     * keyframe gate exists for). False on an audio leg, where the shed ends on
     * the first buffer inside budget.
     */
    keyframeAligned: boolean;
}

/**
 * The `backlogShed` runner config for a clock-paced consumer leg, or
 * `undefined` when the contract is off.
 *
 * `undefined` is the whole legacy story: with `MR_TIME_SYNC_CONTRACT=0` no
 * module sends this config, the runner arms nothing, and the leg behaves
 * exactly as it did — which it must, because a `sync=false` sink drains its own
 * backlog and has no ratchet to guard against.
 */
export function backlogShedConfig(
    services: BacklogShedServices | null | undefined,
    opts: BacklogShedOptions,
): BacklogShedConfig | undefined {
    if (services?.timeSyncContract !== true) return undefined;
    return {
        element: opts.element,
        sink: opts.sink,
        keyframeAligned: opts.keyframeAligned,
        toleranceMs: BACKLOG_SHED_TOLERANCE_MS,
        holdMs: BACKLOG_SHED_HOLD_MS,
        cooldownMs: BACKLOG_SHED_COOLDOWN_MS,
        sanityMs: BACKLOG_SHED_SANITY_MS,
    };
}

File: packages/engine/src/child-process/backlog_shed.py
"""Retained-backlog policy for a clock-paced consumer leg (pure logic, no GStreamer).

THE FAILURE THIS EXISTS FOR — the time-sync contract's latency RATCHET.

Under the contract every presentation consumer runs `sync=true`, so its sink
drains at exactly media rate. That makes backlog one-way. Any downstream hiccup
(a decoder stall, a compositor hitch, a CMA allocation) parks data in the leg's
leaky queues, and a media-rate sink never gives it back: the retained latency
that hiccup created is still there an hour later, and the next hiccup adds to
it. The legacy `sync=false` sink had no such problem — it presented on arrival,
so it gulped any backlog at max speed and drained itself, which is the (now
falsified) assumption the video leg's queue sizing was written against.

Field, 10.9.1.42 (Pi 400, 2026-08-13/14): retained latency climbed over ~16 h
until buffers reached `videoconvert`/`waylandsink` late enough for QoS to drop
almost all of them — decoder still running at 50 fps (codec IRQ ~100/s), glass
at 2.5 fps (`vc4 crtc` IRQ 2.2/s). A live `ts-offset` bump of +1 s restored
60 fps instantly and reverting it put the box straight back to 2.5 fps, which is
what proves the fault is retained SCHEDULE, not throughput.

WHAT THIS DECIDES, and what it deliberately does not. This class owns only the
question "is the leg holding more latency than the route's playout budget D, and
is it time to give it back?" — sustained-excess detection plus rate limiting.
The measuring and the actual dropping live in `gst-pipeline-runner.py`
(`_start_backlog_shedder`), which is where the pads are. Keeping the decision
here makes it testable without a pipeline, the same split as `render_lag.py`.

THE ARITHMETIC. The runner feeds one LATENESS sample per buffer at the shed
point: `lateness = now_running_time - (buffer_running_time + ts_offset)`, i.e.
how far past its scheduled playout slot the buffer is. `ts_offset` IS the
route's playout offset D, so lateness is the excess over budget directly:

    retained pipeline latency  =  lateness + D

Healthy is NEGATIVE (a buffer reaches the shed point before its slot, and the
sink waits); positive is retained backlog beyond D and never recovers on its own.

WHY A FLOOR, NOT AN AVERAGE. Lateness spikes during a stall and relaxes; what
matters is the level it relaxes back TO, because that is buffering the pipeline
will never return. So the trigger is "EVERY sample for `hold_ms` was above
tolerance" — one sample at or below tolerance resets the streak. That is the
per-interval minimum the ratchet reproduction measures, expressed as a rule.

OSCILLATION IS IMPOSSIBLE BY CONSTRUCTION. A shed always ends with lateness at
or below zero (the runner drops until it is), so the streak has to be rebuilt
from scratch: `tolerance_ms` of fresh retention, sustained `hold_ms`, and no
sooner than `cooldown_ms` after the last shed finished. The worst case is one
shed per cooldown, which the events make countable.
"""

# Defaults. Chosen against the video leg's own numbers: the ES queue holds 1 s
# and the jitter queue `bufferMs` (200 ms default), so 250 ms of retention over
# budget is well clear of one absorbed IDR burst (~200 ms of stream time on a
# Pi 4 at 8 Mbps) yet far below `waylandsink max-lateness=1000000000`, the cliff
# past which the sink drops nearly every frame.
DEFAULT_TOLERANCE_MS = 250.0
# Sustained means sustained: 5 s is long enough that a burst being absorbed and
# handed back cannot trip it, short enough that a genuine step (which never
# comes back) is answered in seconds rather than after it has cost frames.
DEFAULT_HOLD_MS = 5_000.0
# One shed per minute per leg, worst case. A shed returns ALL of the excess, so
# frequency buys nothing — and each one costs the video leg the frames up to the
# next IRAP, which the operator sees.
DEFAULT_COOLDOWN_MS = 60_000.0
# Above this, the reading is NOT a backlog. A real retained backlog is bounded
# by the leg's queues (1 s ES + up to 5 s jitter); a sample of tens of seconds
# means the buffer timeline and the pipeline clock are not the same timeline at
# all (an unstamped producer, a segment this code did not expect), and shedding
# on it would drop the entire stream for ever chasing a target it can never
# reach. So it is reported, never acted on. 10 s is `MAX_PLAYOUT_OFFSET_MS`.
DEFAULT_SANITY_MS = 10_000.0


class BacklogShedPolicy:
    """Decides WHEN a clock-paced leg must hand its retained backlog back.

    `observe(lateness_ms, now_ms)` takes one sample per buffer and returns:

        None            nothing to do
        "shed"          start shedding now (returned once per episode)
        "implausible"   the sample is past `sanity_ms` — reported once per
                        episode so the runner can log it; never a shed

    `now_ms` is any monotonic millisecond count; the runner passes the pipeline
    clock's running time, so the policy and the measurement share one time base.
    """

    def __init__(self, tolerance_ms=DEFAULT_TOLERANCE_MS, hold_ms=DEFAULT_HOLD_MS,
                 cooldown_ms=DEFAULT_COOLDOWN_MS, sanity_ms=DEFAULT_SANITY_MS):
        self.tolerance_ms = float(tolerance_ms)
        self.hold_ms = float(hold_ms)
        self.cooldown_ms = float(cooldown_ms)
        self.sanity_ms = float(sanity_ms)
        self.sheds = 0
        self._above_since = None      # start of the current unbroken excess run
        self._last_shed_end = None    # cooldown anchor; None = never shed
        self._implausible = False     # latched so it is reported once, not per buffer

    def reset(self):
        """Drop the streak (not the counters): a flush/re-anchor makes the
        samples either side of it incomparable."""
        self._above_since = None
        self._implausible = False

    def observe(self, lateness_ms, now_ms):
        if lateness_ms is None or lateness_ms != lateness_ms:   # NaN
            return None
        if abs(lateness_ms) > self.sanity_ms:
            # Not a backlog — see DEFAULT_SANITY_MS. The streak is dropped too:
            # a timeline mismatch must never accumulate toward a shed.
            self._above_since = None
            if self._implausible:
                return None
            self._implausible = True
            return "implausible"
        self._implausible = False
        if lateness_ms <= self.tolerance_ms:
            self._above_since = None
            return None
        if self._above_since is None:
            self._above_since = now_ms
            return None
        if now_ms - self._above_since < self.hold_ms:
            return None
        # Sustained excess. The cooldown gate does NOT reset the streak: a leg
        # that is still over budget when the cooldown expires sheds immediately,
        # rather than paying the hold window again.
        if self._last_shed_end is not None and now_ms - self._last_shed_end < self.cooldown_ms:
            return None
        return "shed"

    def shed_finished(self, now_ms):
        """The runner reached its target (or gave up). Arms the cooldown and
        forces the streak to be rebuilt from scratch."""
        self.sheds += 1
        self._last_shed_end = now_ms
        self._above_since = None
